### ISSUE TITLE
Save metadata to tree sequences produced by the msprime backend

### ISSUE THREAD

**Issue by bodkan** [bodkan/slendr#79]
**Background**:

Until now, everything in slendr assumes SLiM tree sequences and all R/_tskit_ interface functions do loading and processing with pyslim. Now that [_slendr_ models can be executed by a built-in _msprime_ backend without any change](https://www.slendr.net/articles/vignette-07-backends.html), I have started working on generalizing the `ts_load()`, `ts_mutate()`, `ts_<statistic>()` tskit-interfaces of slendr so that they can handle msprime tree sequences as well. At this point, loading and processing is done by pyslim and assumes that a tree sequence is `SlimTreeSequence`.

For all of this to work, I need to save various pieces of metadata with the tree sequence so that `ts_load()` and friends in _slendr_ know how to interpret the output in the _slendr_ context. With our SLiM backend this is extremely easy: I first [construct an Eidos `Dictionary` with all the relevant pieces](https://github.com/bodkan/slendr/blob/11c35d32ded99897ad446ecfcdd54d3cf3751bc4/inst/extdata/script.slim#L594) and save that along the tree sequence by supplying the dictionary of dictionaries when calling [`sim.treeSeqOutput()`](https://github.com/bodkan/slendr/blob/11c35d32ded99897ad446ecfcdd54d3cf3751bc4/inst/extdata/script.slim#L419). So far so good.

**Problem**:

I'm having massive issues doing the same thing on the [msprime backend](https://github.com/bodkan/slendr/blob/main/inst/extdata/script.py). :(

I read the [tutorial](https://tskit.dev/tutorials/metadata.html#modifying-metadata-and-schemas) a bunch of times, trying to understand what is what (there are some crashed code chunks but I think I get the overall picture) as well as the [other pieces of docs](https://tskit.dev/tskit/docs/stable/metadata.html#sec-metadata)... but I have not managed to save even a single string value to the output tree sequence.

I'm rarely this lost with anything related to programming so I'm probably missing some crucial piece of information to understand how metadata & codecs & JSON are done in msprime and how to save them to the top-level. Most importantly, I did not find anything close to how SLiM can save a dictionary of dictionaries automatically in a straightforward way.

**Question:**

@petrelharp You mentioned at some point you have started saving various metadata bits (parameters used, etc.) with all your tree sequences. In fact, this is what inspired me to do [exactly the same thing](https://github.com/bodkan/slendr/blob/11c35d32ded99897ad446ecfcdd54d3cf3751bc4/inst/extdata/script.slim#L594) in slendr for the SLiM backend. Do you perhaps have a working example in msprime that I could study?

I'm not looking to do anything complex. I just want to save the same dict-of-dicts metadata (slendr version, simulation parameters, table with times/names/populations of individuals being recorded, nothing else), exactly the same thing as we're currently doing on the [SLiM side](https://github.com/bodkan/slendr/blob/11c35d32ded99897ad446ecfcdd54d3cf3751bc4/inst/extdata/script.slim#L594).

Also pinging @bhaller in case he has some advice from his work on metadata encoding in SLiM. I know things are very different in msprime, but as I said, I must be missing something completely obvious. (Sorry for the spam, but I am already desperate :)).

---------------------

Here is a sketch of what I'm trying to do in msprime (and a mirror of what we currently do on the SLiM side of things):

```
# Let's say that this is a version of slendr metadata Dictionary, similar to what
# we save in SLiM
slendr_metadata = {
    "slendr" : {
        "version" : "0.9",
        "backend" : "msprime",
        "description" : "This is a fancy slendr model",
        "sampling" : {
            1 : { "name" : "ind_1", "pop" : "pop1", "time" : "30000"},
            2 : { "name" : "ind_2", "pop" : "pop1", "time" : "20000"},
            3 : { "name" : "ind_3", "pop" : "pop1", "time" : "10000"}
        },
        "arguments" : {
            "SEQUENCE_LENGTH" : 1000000,
            "RECOMB_RATE" : 1e-8,
            "blah" : "blah blah blah"
        }
    }
}

# according to docs, this apparently skips some internal validation which
# seems OK for a toy example case
ts.tables.metadata_schema = tskit.MetadataSchema({"codec": "json"})

# the following command gives:
# TypeError: If no encoding is set metadata should be bytes, found <class 'dict'>
ts.tables.metadata = slendr_metadata

# inspired by an unrelated (but related to encoding and bytes and what not) piece of
# info elsewhere in tskit docs, this doesn't give error...
import pickle
ts.tables.metadata = pickle.dumps(slendr_metadata)
ts.dump("new.trees")

# (attempt to do this directly on ts.metata fails with "can't set attribute")

# ... but when I run the following and load the tree sequence again, I see no metadata

# ipdb> new_ts = tskit.load("new.trees")
# ipdb> new_ts.metadata
# b''
# ipdb> new_ts.tables.metadata
# b''
```

I do acknowledge that my attempt is probably broken on multiple levels (again, codecs, bytes, encodings, JSON, what? 😰) but I hope that it demonstrates what I was hoping to do, that would approximate how things work in SLiM in this context.

Also, I'm using the latest releases of tskit and msprime.

**Comment by bhaller** [comment]
Well, I have never done this with msprime, so I don't know much.  I'll tag @benjeffery here, though, and I'll take a few stabs in the dark.  My impression is that tskit provides no help for assembling or parsing JSON strings; since it is pure C there's no JSON dictionary class or anything like that, as there is in Eidos.  So my guess is that you will want to choose a JSON library that you like, for whatever language you're working in, and use it to turn whatever dictionary-ish object you have into a JSON string.  Then you just hand that JSON string to tskit as a pointer to char and a number of bytes.  That's what SLiM/Eidos do internally, using the nlohmann C++ JSON library (https://github.com/nlohmann/json).  You would do the same on the way back out: get the pointer to char and number of bytes from tskit, and use that to make a JSON string that you hand to the JSON library of your choice, which turns it back into a dictionary-ish object for you.

My other thought is that there may well be example code for doing this in Python already, at least in the form of internal test code that Ben and others have probably written to test out tskit's metadata support.  So you might dig around inside the test code for the python wrappers for tskit/msprime to see what you find there.  Ben can probably point you to exactly what you want to look at, though.

I hope this helps, and I apologize if any of the above turns out to be misinformation!  :->

**Comment by bodkan** [comment]
Thanks a lot, Ben! This was extremely useful. I hope I'll be able to do similar JSON shenanigans with pure Python, because I'm too soft for _"pointer to char and a number of bytes"_ type of work at this point! :) I'll wait if Peter or Ben will be able to help. In the meantime, crawling through tskit's metadata-related tests is an excellent idea! 👍 What I need to do is not rocket science, so there must be a simple answer somewhere.

**Comment by clwgg** [comment]
Hi @bodkan, sorry for randomly jumping in here, but the issue you are encountering is not directly related to the metadata or how tskit handles json, but due to the fact that the `tables` object is not writable if accessed through its `tree_sequence` parent object. So your issue should be resolved if you do:

``` python
import msprime
import tskit

ts = msprime.sim_ancestry(
    samples=2,
)

slendr_metadata = {
    "slendr": {
        "version": "0.9",
        "backend": "msprime",
        "description": "This is a fancy slendr model",
        "sampling": {
            1: {"name": "ind_1", "pop": "pop1", "time": "30000"},
            2: {"name": "ind_2", "pop": "pop1", "time": "20000"},
            3: {"name": "ind_3", "pop": "pop1", "time": "10000"},
        },
        "arguments": {
            "SEQUENCE_LENGTH": 1000000,
            "RECOMB_RATE": 1e-8,
            "blah": "blah blah blah",
        },
    }
}

tables = ts.tables
tables.metadata_schema = tskit.MetadataSchema({"codec": "json"})
tables.metadata = slendr_metadata
ts = tables.tree_sequence()
ts.dump("new.trees")
```
It's probably not ideal that this happens silently and there is active discussion on the tskit site to resolve that, see e.g. https://github.com/tskit-dev/tskit/issues/1906

**Comment by bodkan** [comment]
Wow, thank you so much, @clwgg! This worked perfectly!

I can't believe I was _this_ close, yet at the same time _this_ away from a working solution!

In the metadata tutorial, I read

> This [MetadataSchema](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.MetadataSchema) can then be assigned to a table or the top-level tree sequence e.g. [metadata_schema](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.IndividualTable.metadata_schema):
> ```
> tables = tskit.TableCollection(sequence_length=1)  # make a new, empty set of tables
> tables.individuals.metadata_schema = complex_schema
> ```

But that's when I got stuck -- I didn't want to create new empty set of tables, I just wanted to _add_ information to metadata already present! But my brain didn't get past this, because at that point I was already entangled in trying to figure out encoding/decoding etc., as I assumed it's probably that where I'm making a mistake. However, it was the bit with `ts.dump_tables()` was exactly what I was missing. Super obvious in hindsight.

Thank you again for jumping in and for helping me out. You saved me a ton of time!

Now I should have everything to unify the interface for loading/processing/analysing msprime and SLiM tree sequences in _slendr_. 🚀

**Comment by bodkan** [comment]
EDIT (for anyone who ends up here while searching for a solution to a similar issue):

Oh man, it's actually spelled out literally in the [API manual](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence):

> TreeSequences are immutable. To change the data held in a particular tree sequence, first get the table information as a [TableCollection](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TableCollection) instance (using [dump_tables()](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.dump_tables)), edit those tables using the [tables api](https://tskit.dev/tskit/docs/stable/python-api.html#sec-tables-api), and create a new tree sequence using [TableCollection.tree_sequence()](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TableCollection.tree_sequence).

Unfortunately, I got stuck even reading the tutorials so re-reading the API afterwards didn't seem like something that would help (WRONG 😕).

Well, at least now I understand a bit better how the tables are internally organized and stored (which is a useful thing to know regardless of what I do).

**Comment by benjeffery** [comment]
@bodkan Sorry, I didn't see this until now. I wrote the metadata docs - and I think they could be improved to assume less knowledge about the general tskit API or at least point the way to what should be read first. Glad you sorted your issue, never be shy about making an issue or discussion directly over at the tskit repo!

**Comment by bodkan** [comment]
Hello @benjeffery! No worries. I didn't have to wait at all as my friends above jumped to the rescue immediately. :)

The docs you wrote are excellent and very detailed. I hope what I wrote didn't read like criticism and I'm sorry if it did. You all have done an outstanding job with the documentation and tutorials considering how massive the tskit framework is. Plugging tree sequence capabilities into my slendr R package was an absolute pleasure.

My issue as a complete newbie was that I was missing a super minimal working example to get started. I think I approached the problem from the wrong end and, as a result, built and extremely overcomplicated mental model of what probably needs to be done in order to accomplish what I needed (codecs, schemas, etc.).

Once I got help [here](https://github.com/bodkan/slendr/issues/79#issuecomment-1030501840) with the simple snippet of code, I was able to implement what I needed without any issues and the solution was, indeed, completely trivial. So perhaps if anything, some simple example like that might be a useful addition to the tutorial, at least for people who get easily confused as me. :)
